### ISSUE TITLE
multiple metavars in one line with string concatination fails to match

### ISSUE THREAD

**Issue by ajinabraham** [semgrep/semgrep#514]
https://sgrep.live/Go2?version=experimental

Why Line 5 is not getting matched?
Multiple metavars in one line is not respected.

On a side note. A good thing to have is the **ability to check if a metavar in any form reach a sink**.  For example in this case, 
```
$APP.$METHOD(..., function $FUNC($REQ, $RES, ...) {
              ...
          });
```

I need to know if the metavar `$REQ` reaches `connection.query()` in any format. It can be multiple variables, string concatenation, array, object ($REQ.$FOO, $REQ.$FOO.$BAR,  $REQ.$FOO('123'),  $REQ.$FOO['123']) etc. 

sgrep will be super useful if **that kind of inspection feature** can be added. With current implementation. I have to write n permutation combinations of string matches like prepending, appending, in between join, and multiple occurrences of $REQ object. Writing a rule like that will always have something missing in it and reduce detection capabilities.

**Comment by aryx** [comment]
@ajinabraham we are working on a specialised version of the yaml file for taint tracking.
Here is an example: https://github.com/returntocorp/semgrep/commit/835adfd8d0f0163071bd86bb6361d456d92ea7da#diff-57664a5cb0ff7050443755f7ce537f10
It's partly implemented right now but feel free to give some feedback of where you would like this to go.
The idea is that you use semgrep patterns to specify the shape of the source, the sink, and sanitizers, and semgrep internally does all the taint tracking using dataflow. It's just intraprocedural right now.

**Comment by aryx** [comment]
There's also the deep-expression operator <... ...>
You simply do connection.query(<... $REQ ...>, ...)
and it will match when $REQ appears anywhere inside the first argument of connection.query
You can use this operator anywhere where an expression is expected, e.g.,
if (<... true ...>)
to detect places where true is used inside a conditional.

**Comment by aryx** [comment]
And the bug is because you use $VAR2 which is not recognized as a metavariable ...
I'll change that.

**Comment by aryx** [comment]
If you change $VAR2 with $VARR it would work (but we want $VAR2 to work too indeed)

**Comment by DrewDennison** [comment]
@aryx thanks for the clarification on the $VAR2 I'm going to close this as resolved. @ajinabraham I confirmed this works in https://semgrep.live/36g

**Comment by ajinabraham** [comment]
@aryx Taint analysis is super useful!
I like <..$REQ..> saves me a tone of lines for writing a rule. Do you have this feature available in experimental or develop or in 0.5.0?

**Comment by ajinabraham** [comment]
Nevermind I saw the release log.
